Radiusю

HelenMamontova opened the pull request:
Search for parameters auth and send, get value of the parameter send added.

HelenMamontova committed:
Radius. Header file <mutex>, <jthread.hpp> added. Functions start, stop
declaration changed. Class members m_thread, m_mutex added. Function Run
declaration added.

HelenMamontova committed:
Radius. Header files added. Functions Start, Stop, Run definition added.

HelenMamontova committed:
Radius. New files server.h, server.cpp added.

HelenMamontova committed:
Radius. Definition of functions startReceive, makeResponse, handleSend,
handleReceive added.

HelenMamontova committed:
The Run function name fixed. Definition of secret variable sdded.

HelenMamontova committed:
The file server.cpp added to add_library command, command
find_package(OpenSSL 1.0.0 Required) added, OpenSSL::Crypto added to
target_link_libraries command in the block if(BUILD_MOD_RADIUS).

HelenMamontova committed:
Parameter token added to function Run() declaration.

HelenMamontova committed:
Radius. Method Start: variable isRunning=true removed, m_thread
definition changed. Method Stop: variable isRunning=false removed,
m_thread.joinable check added, isRunning check added, request_stop call added.
Method Run: parameter token added, variables secret, port removed,
object lock and isRunning=true added before cycle while,
cycle while added, isRunning=false added after cycle while.

HelenMamontova committed:
Radius. Hold the mutex removed,extra conditions
for m_thread.join() removed in the Stop function.

HelenMamontova committed:
Radius. Cycle while removed in function Run.

HelenMamontova committed:
Radius. The variables isRunning, errorStr replaced by m_running,
m_errorStr in class Radius.

HelenMamontova committed:
Radius. The variable isRunning replaced by m_running in function Run.

HelenMamontova committed:
Radius. Object name io_service changed to ioService in the function Run.

HelenMamontova committed:
Method SetRunning declaration added to the class RADIUS.

HelenMamontova committed:
Method SetRunning definition added. Method SetRunning call added to
function Run.

HelenMamontova committed:
Radius. Namespace STG added.

HelenMamontova committed:
Radius. Declaration using STG::Server added.

HelenMamontova committed:
Radius. Header file "stg/logger.h" added. RADIUS class member m_loger added.

HelenMamontova committed:
Radius. Header file "stg/common.h" added. Initialization of m_logger
added to constructor RADIUS. Output cerr replaced by logger and
printfd().

HelenMamontova committed:
Radius. Variable except added, messages fixed in the function Run.

HelenMamontova committed:
Radius. Thread join logic fixed in the function Stop.

HelenMamontova committed:
Keyword const added to std::lock_guard in the function SetRunning.

HelenMamontova committed:
Radius. Class member m_mutex moved to the top of list.

HelenMamontova committed:
Radius. Class member m_logger put after the m_running.

HelenMamontova committed:
Radius. Unnecessary variable except removed in the function Run.

HelenMamontova committed:
Radius. Extra whitespace removed in the function Stop. Function
IsRunning definition added.

HelenMamontova committed:
Radius. Function IsRunning declaration changed.

HelenMamontova committed:
Radius. Formatting fixed.

HelenMamontova committed:
Radius. Extra symbols '//' removed.

HelenMamontova committed:
Radius. Class member m_running initialization added to constructor
RADIUS.

HelenMamontova committed:
Radius.Header file "stg/module_settings.h" added. Class members variable
m_settings and method SetSettings added to class RADIUS.

HelenMamontova committed:
Radius. The parameters Secret, Dictionaries added, parameter Port
changed, other parameters removed.

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Radius. Header files added. Class RAD_SETTINGS added. Methods
SetSettings, copy constructor RADIUS, assignment operator added. Method
ParseSettings changed. Formatting fixed.

HelenMamontova committed:
Radius. Parameter default changed to optional and parameters value
commented out for the parameters dictionaries, port.

HelenMamontova committed:
Radius. Description of parameters secret, port, dictionaries changed.

HelenMamontova committed:
Radius. Copy constructor RADIUS and assignment operator declaration
changed and moved to public section.

HelenMamontova committed:
Radius. Class member m_mutex moved to the top of section private.

HelenMamontova committed:
Radius. The m_port variable value initialization changed to 1812,
m_dictionaries variable initialization added to constructor RAD_SETTINGS.

HelenMamontova committed:
Radius. Check for missing parameters port and dictionaries removed in function ParseSettings.

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Radius. The dist directory added.

HelenMamontova committed:
Radius. The methods declaration stop and start added.

HelenMamontova committed:
Radius. The methods stop, start definition added. Method start call
added to Server constructor.

HelenMamontova committed:
Radius. The header files "server.h", <boost/asio.hpp>, <memory> added.
The class members m_ioService, m_server added.

HelenMamontova committed:
Radius. The header files "server", <boost/asio.hpp>, <string> removed.
The method m_server->stop() call added to method Stop.

HelenMamontova committed:
Radius. Header files <stop_token> and <jthread.hpp> added. Parameter
token added to constructor Server. Class member m_token added.

HelenMamontova committed:
Radius. Parameter token and m_token class member initialization added to Server
constructor definition. The check m_token.stop_requested() added to
method handleSend before startReceive method call.

HelenMamontova committed:
Radius. The m_server class member initialization removed.

HelenMamontova committed:
Radius. The m_server->stop() call adnd m_thread.request_stop() call
reodered, the m_server check added before m_server->stop call in the
method Stop. Parameter token added to constructor Server in the function
Run.

HelenMamontova committed:
Radius. Header file "stg/logger.h" added. Class member m_logger added to
class Server.

HelenMamontova committed:
Radius. Header file "stg/common.h" added, header file <iostream>
removed. Class member m_logger initialization added to constructor
Server. Object std::cout replaced by function printfd and m_logger.

HelenMamontova committed:
Radius. Class member m_logger initialization changed in the constructor
Server. Check token moved before checking the error code in the
handleSend method.

HelenMamontova committed:
Radius. Check token added to handleReceive method.

HelenMamontova committed:
Radius. Header file <stop_token> replaced by <stop_token.hpp>, header
file <jthread.hpp> removed.

HelenMamontova committed:
Radius. Parameter m_logger added to constructor Server call in the
method Run.

HelenMamontova committed:
Radius. Parameter logger added to constructor Server declaration. Class
member m_logger added to class Server.

HelenMamontova committed:
Radius. Parameter logger added to constructor Server definition. Class
member m_logger initialization changed in the constructor Server.

HelenMamontova committed:
Radius. Extra whitespace removed ib the method Run.

HelenMamontova committed:
Radius. The order of operations changed in the method handleSend.

HelenMamontova committed:
Radius. The order of operations changed in the method handleReceive.

HelenMamontova committed:
Radius. Operator return removed when checking ec, operator else removed
when checking packet in the handleReceive method.

HelenMamontova committed:
Radius. Parameter token replaced by std::move(token) in the constructor Server
in method Run.

HelenMamontova committed:
Radius. Class member m_token initialization value changed to
std::move(token) in the constructor Server.

HelenMamontova committed:
Radius. The c_str() function added to message when checking error in the
handleSend and handleReceive methods.

HelenMamontova committed:
Radius. Header file <cstring> added.

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Radius. Header files "stg/users.h", "stg/user.h", "stg/user_property.h"
added. The using statements added. The findUser, SetUsers functions
declaration added. Class member m_users added.

HelenMamontova committed:
Radius. Header files "radproto/attribute_types.h", <cstdint> added.
The m_users class member initialization added to constructor Server. The
findUser function definition added. The findUser function call added to
makeResponse function.

HelenMamontova committed:
Radius. Getting the values of login and password fixed in the function
findUser. Extra functions printfd call added.

HelenMamontova committed:
Radius. Using statements added. Method SetUsers added, variable class
member m_users added to class RADIUS.

HelenMamontova committed:
Radius. The class member variable m_users initialization added to
constructor Radius. Parameter m_users added to constructor Server call
in the method Run.

HelenMamontova committed:
Radius. Parameter users added to constructir Server declaration.
Function SetUsers removed fron class Server.

HelenMamontova committed:
Radius. Parameter users added to constructor Server definition. Variable
class member m_users initialization added to constructor Server.

HelenMamontova committed:
Radius. Function c_str added to login and password in m_logger and
printfd function in the function findUser.

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Radius. Extra functions printfd removed.

HelenMamontova committed:
Radius. The using declaration for UserPtr, ConstUserPtr removed.

HelenMamontova committed:
Radius. Looking for user moved after check request type in the
makeResponse function.

HelenMamontova committed:
Radius. The using declaration for UserPtr, ConstUserPtr removed. Header
files "stg/user.h", "stg/users.h", "stg/user.property.h" removed.

HelenMamontova committed:
Radius. Header files "stg/user.h", "stg/users.h" added.

HelenMamontova committed:
Radius. Function findUser moved to private.

HelenMamontova committed:
Radius. Return type of function findUser changed to bool.

HelenMamontova committed:
Radius. Return type of function findUser definition and call changed to bool.

HelenMamontova committed:
Radius. Variables attributesEmpty, vendorSpecificEmpty added for packet
ACCESS_REJECT in the function makeResponse.

HelenMamontova committed:
Radius. Type STG::Users* replaced by Users* in parameter users in the
constructor Server definition.

HelenMamontova committed:
Radius. Condition and code order changed when packet is returned in
makeResponse function.

HelenMamontova committed:
Radius. Unnecessary variables attributeEmpty, vendorSpecificEmpty
replaced by {} when ACCESS_REJECT packet is returned in makeResponse
function.

HelenMamontova committed:
Radius. Prefix STG of user object removed, user is initialized by
nullptr and moved to the point before call FindByName function in the
findUser function.

HelenMamontova committed:
Radius. Extra printfd function call removed in the function findUser.

HelenMamontova committed:
Radius. The way of initializing the object user changed in the findUser
function. Error log removed in the makeResponse function.

HelenMamontova committed:
Radius. Prefix STG of class member m_users and of parameter u in
SetUsers function removed.

HelenMamontova committed:
Radius. Prefix STG of class member m_users and of parameter users
in the constructor Server declaration removed.

HelenMamontova committed:
Radius. The messages of m_logger and printfd changed in function
findUser.

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Radius. Search for parameters auth and send added to the ParseSettings function.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -46,14 +46,27 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
+
+    pv.param = "auth";
+    pvi = std::find(s.moduleParams.begin(), s.moduleParams.end(), pv);
+    if (pvi != s.moduleParams.end() )
```
Extra whitespace at the end.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -46,14 +46,27 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
+    if (pvi != s.moduleParams.end() )
+    {
+        pv.param = "send";
+        std::string sendValue;
```
You do not need this variable, see below.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -46,14 +46,27 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
+    {
+        pv.param = "send";
+        std::string sendValue;
+        std::vector<ParamValue> sectionsParams = pvi->sections;
```
You do not need to store sections in the additional variable, see below.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -46,14 +46,27 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
+        pv.param = "send";
+        std::string sendValue;
+        std::vector<ParamValue> sectionsParams = pvi->sections;
+        auto pva = std::find(sectionsParams.begin(), sectionsParams.end(), pv);
```
Use `sections` directly:
```c++
auto pva = std::find(pvi->sections.begin(), pvi->sections.end(), pv);
```

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -46,14 +46,27 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
+        if (pva != sectionsParams.end() && !pva->value.empty())
+        {
+            sendValue = pva->value[0];
+            printfd(__FILE__, "ParseSettings Value of send: '%s'\n", sendValue.c_str());
```
Consider `pva->value[0].c_str()` instead of `sendValue`.

HelenMamontova committed:
Radius. Extra whitespace removed, unnecessary variables sendValue and
sectionsParams removed in ParseSettings function.

HelenMamontova committed:
Radius. The struct AttrValue added.

HelenMamontova committed:
Radius. Header files <boost/tokenizer.hpp>, <vector>, <utility> added.
Parsing attributes of field send in section auth of mod_radius.conf
added to the function ParseSettings.

HelenMamontova committed:
Radius. Filling vector of pairs attrSend added. Extra printfd function
removed.

HelenMamontova committed:
Radius. The Sign enumeration elements names changed in the struct
AttrValue.

HelenMamontova committed:
Radius. Using declaration for struct AttrValue added. Function ParseSendAttr definition added
and call in the function ParseSettings added.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -20,11 +20,23 @@ namespace STG
 
+    struct AttrValue
+    {
+        enum class Sign
```
Let's call it `Type`.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -20,11 +20,23 @@ namespace STG
+    {
+        enum class Sign
+        {
+            NOT_VALUE,
```
Here would be `PARAM_NAME`

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -20,11 +20,23 @@ namespace STG
+        enum class Sign
+        {
+            NOT_VALUE,
+            IS_VALUE
```
And here would be `VALUE`.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -20,11 +20,23 @@ namespace STG
 
     class Users;
 
+    struct AttrValue
```
Let's put it into the `RAD_SETTINGS` class.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,19 +1,64 @@
+
+    for (const auto& token : tokens)
+    {
+        boost::char_separator<char> sp(" =");
```
What is there there is no a whitespace before the '='?

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,19 +1,64 @@
+        for (const auto& t : tok)
+            parsedSendAttr.push_back(t);
+
+        std::string key = parsedSendAttr[0];
```
What if `parsedSendAttr` is empty?

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,19 +1,64 @@
+            parsedSendAttr.push_back(t);
+
+        std::string key = parsedSendAttr[0];
+        std::string valueName = parsedSendAttr[1];
```
What if `parsedSendAttr` has only one element?

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,19 +1,64 @@
+        if (valueName[0] == '\'')
+        {
+            valueName.erase(0, 1);
+            valueName.erase(valueName.length() - 1, 1);
```
What if the last symbol of `valueName` is not a quotation mark?

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,19 +1,64 @@
+
+            attrValue.value = valueName;
+            attrValue.sign = AttrValue::Sign::IS_VALUE;
+            keyValuePairs.emplace_back(key, attrValue);
```
Consider `keyValuePairs.emplace_back(key, AttrValue{valueName, AttrValue::Sign::IS_VALUE})`. Same below.

HelenMamontova committed:
Radius. The enum class Sign renamed to Type, object sign renamed to
type.

HelenMamontova committed:
Radius. The enum class Sign renamed to Type, object sign renamed to type
in ParseSendAttr function and ParseSettings function.

HelenMamontova committed:
Radius. The enumerator names NOT_VALUE changed to PARAN_NAME, IS_VALUE
changed to VALUE in enum Type.

HelenMamontova committed:
Radius. Radius. The enumerator names NOT_VALUE changed to PARAM_NAME, IS_VALUE changed to VALUE in
ParseSendAttr function.

HelenMamontova committed:
Radius. The AttrValue struct moved into the RAD_SETTINGS class.

HelenMamontova committed:
Radius. The using statement for struct AttrValue removed. The using alias for
struct AttrValue added.

HelenMamontova committed:
Radius. The vector parsedSendAttr check added to the function
ParseSendAttr.

HelenMamontova committed:
Radius. The ParseSendAttr method added, m_logger variable class member
added to RAD_SETTINGS class.

HelenMamontova committed:
Radius. ParseSendAttr function changed to method of RAD_SETTINGS class.
Vector parsedSendAttr check for emptiness and check vector size added to
the ParseSendAttr method. Class member m_logger initialization added to
initialization list to RAD_SETTINGS constructor.

HelenMamontova committed:
Radius. Quotes check for the ValueName added to the ParseSendAttr
method.

HelenMamontova committed:
Radius. Extra variables attrValue.value and attrValue.type removed in
the ParseSendAttr method.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,82 @@
+
+    tokenizer tokens(fieldSendAttr, sep);
+
+    AttrValue attrValue;
```
Unused variable.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,82 @@
+        for (const auto& t : tok)
+            parsedSendAttr.push_back(t);
+
+        if (parsedSendAttr.empty())
```
Consider the following:
```c++
if (parsedSendAttr.size() != 2)
{
    m_logger("Send attribute specification has an incorrect format: '%s'.", token.c_str());
    ...
    return {};
}

auto type = AttrValue::Type::PARAM_NAME;
if (/*check if parsedSendAttr[1] has quotes*/)
{
    type = AttrValue::Type::Value;
    /*strip quotes from the parsedSendAttr[1]*/
}
else if (/*check if the quotes are unpaired*/)
{
    /*log error*/
    return {};
}
keyValuePairs.emplace_back(parsedSendAttr[0], AttrValue{parsedSendAttr[1], type});
```

HelenMamontova committed:
Radius. Extra variable attrValue removed in ParseSendAttr method.

HelenMamontova committed:
Radius. Send attributes parsing changed in method ParseSendAttr.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,71 @@
     return &plugin;
 }
 
+std::vector<std::pair<std::string, AttrValue>> RAD_SETTINGS::ParseSendAttr(std::string fieldSendAttr)
```
First, `const std::string&`

Second, let's call the argument `val` or `value`.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,71 @@
+
+    tokenizer tokens(fieldSendAttr, sep);
+
+    std::vector<std::pair<std::string, AttrValue>> keyValuePairs;
```
Let's call it `res`. It's the result of the function.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,71 @@
+        boost::char_separator<char> sp(" =");
+        tokenizer tok(token, sp);
+
+        std::vector<std::string> parsedSendAttr;
```
Let's call it `kv` or `keyValue`.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,71 @@
+
+        auto type = AttrValue::Type::PARAM_NAME;
+        std::string valueName = parsedSendAttr[1];
+        if (valueName[0] == '\'' && valueName[valueName.length() - 1] == '\'')
```
Consider using `valueName.back()` instead of `valueName[valueName.length() - 1]`. You can also use `front()` for accessing the first char, for symmetry. But it's up to you.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -25,8 +25,21 @@ namespace STG
+
             const std::string & GetStrError() const { return m_errorStr; }
             int ParseSettings(const ModuleSettings & s);
+            std::vector<std::pair<std::string, AttrValue>> ParseSendAttr(std::string fieldSendAttr);
```
Let's make it private.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -46,14 +95,29 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
+        {
+            printfd(__FILE__, "ParseSettings Value of send: '%s'\n", pva->value[0].c_str());
+
+            std::vector<std::pair<std::string, AttrValue>> keyValuePairs = ParseSendAttr(pva->value[0]);
```
This variable must be a `RAD_SETTINGS` private member, with a public accessor function.

HelenMamontova committed:
Radius. Argument fieldSendAttr name changed to value and argument type
changed to const std::string& in ParseSendAttr method. Vector
keyValuePairs name changed to res.

HelenMamontova committed:
Radius. Argument fieldSendAttr name changed to value and argument type
changed to const std::string& in ParseSendAttr method.

HelenMamontova committed:
Radius. Vector name parsedSendAttr changed to keyValue in ParseSendAttr
method.

HelenMamontova committed:
Radius. Functions front() and back() are used to get the first and last
characters of string valueName in ParseSendAttr method.

HelenMamontova committed:
Radius. Method declaration ParseSendAttr moved to private in class
RAD_SETTINGS.

HelenMamontova committed:
Radius. Formatting fixed. Class member m_res and public function
GetRes added to class RAD_SETTINGS.

HelenMamontova committed:
Radius. Extra variable res removed. Class member m_res definition added
to ParseSendAttr function and ParseSettings function.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -18,48 +18,46 @@ extern "C" STG::Plugin* GetPlugin()
             return {};
         }
-        keyValuePairs.emplace_back(parsedSendAttr[0], AttrValue{valueName, type});
+        m_res.emplace_back(keyValue[0], AttrValue{valueName, type});
```
You use a private member `m_res` as a container. And later you return it by value. And the returned value you assign back to the `m_res`.

Here is what happens. First, you fill the `m_res` with parsed data. Now `m_res` contains all necessary information and no further actions are needed. When you return the `m_res` by value, you make a copy of it. And now you have two identical copies of the information. And then you assign the copy back to the original. This assignment does nothing. It's like the following:

```c++
int a = 5;
int b = a;
a = b;
```

One solution would be to return `void`. You've done all the necessary work inside the function, nothing else is needed. However, there are flaws doing thing this way. When the function works, it changes the value of the `m_res`. For now it's nothing, because it is initially empty. However, if we add a reload capability later, it may contain correct data. And you will gradually change it. It will introduce a race condition and you would have to guard each `emplace_back` with a mutex. What's worse, if there is an issue with the input data, the parsing may fail. Buth the original value of the `m_res` would be damaged.

The better way is to use a local container `res`, not a class member. And fill it with values. And return it by value. And in the end, assign the `m_res` to the returned value.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -37,19 +37,22 @@ namespace STG
-            const std::string & GetSecret() const { return m_secret; }
+            const std::string& GetDictionaries() const { return m_dictionaries; }
+            const std::string& GetSecret() const { return m_secret; }
+            const std::vector<std::pair<std::string, AttrValue>>& GetRes() const { return m_res; }
```
`GetRes` and `m_res` are bad names. The mean nothing. This field represents the contents of the `send` field in the config, so let's call it `GetSend` or `GetSendPairs`, and `m_send` or `m_sendPairs` respectively.

HelenMamontova committed:
Radius. Vector res added, variable m_res replaced by res in
ParseSendAttr function.

HelenMamontova committed:
Radius. Function name GetRes changed to GetSendPairs, class member
vaiable name m_res changed to m_sendPairs in class RAD_SETTINGS.

HelenMamontova committed:
Radius. Class member vaiable name m_res changed to m_sendPairs in the
ParseSettings function.

HelenMamontova committed:
Radius. Parameter auth added to mod_radius.conf.

HelenMamontova committed:
Radius. The ParseMatchAttr method declaration and m_matchPairs class
member added to RAD_SETTINGS class.

HelenMamontova committed:
Radius. The ParseMatchAttr function definition added. Match attributes
parsing added to ParseSetting function.

HelenMamontova committed:
Radius. The ParseMatchAttr method removed, the ParseSendAttr method
renamed to ParseAuthAttr in the class RAD_SETTINGS.

HelenMamontova committed:
Radius. The ParseSendAttr method definition and call renamed to
ParseAuthAttr. Parameter paramName added to ParseAuthAttr method. The
ParseMatchAttr method definition removed. ParseMatchAttr method call
replaced by ParseAuthAttr.

madf commented on projects/stargazer/inst/linux/etc/stargazer/conf-available.d/mod_radius.conf:
```diff
@@ -17,4 +17,8 @@
 
+    <auth>
+        send = "Framed-IP-Address = currip, Framed-Protocol = \'PPP\'"
+        match = "Framed-IP-Address = currip, Framed-Protocol = \'PPP\'"
```
This 'match' example does not make much sense. Let's use something like this:
```
match="Calling-Station-Id=userdata0"
```

HelenMamontova committed:
Radius. Attribute for parameter auth/match changed.

HelenMamontova committed:
Radius. Parameter autz added.

HelenMamontova committed:
Radius. ParseAuthAttr function renamed to ParseAuthAutzAttr.

HelenMamontova committed:
Radius. ParseAuthAttr function renamed to ParseAuthAutzAttr. The autz
attributes parsing added, m_logger and printfd messages changed in the
ParseAuthAutzAttr function.

HelenMamontova committed:
Radius. ParseAuthAutzAttr function renamed to ParseSectionsAttr. The MakeKeyValuePairs function
definition added.

HelenMamontova committed:
Radius. ParseAuthAutzAttr function renamed to ParseSectionsAttr. The messages
of m_logger and printfd changed in the ParseSectionsAttr function.
The MakeKeyValuePairs function definition and call added.

madf commented on projects/stargazer/inst/linux/etc/stargazer/conf-available.d/mod_radius.conf:
```diff
@@ -19,6 +19,11 @@
 
     <auth>
         send = "Framed-IP-Address = currip, Framed-Protocol = \'PPP\'"
```
Let's use "Service-Type = 'Framed-User'" instead of "Framed-Protocol = 'PPP'" here.

madf commented on projects/stargazer/inst/linux/etc/stargazer/conf-available.d/mod_radius.conf:
```diff
@@ -19,6 +19,11 @@
     </auth>
+
+    <autz>
+        send = "Framed-IP-Address = currip, Framed-Protocol = \'PPP\'"
```
And use "Service-Type = 'Login-User'" instead of "Framed-Protocol = 'PPP'" here.

madf commented on projects/stargazer/inst/linux/etc/stargazer/conf-available.d/mod_radius.conf:
```diff
@@ -19,6 +19,11 @@
+
+    <autz>
+        send = "Framed-IP-Address = currip, Framed-Protocol = \'PPP\'"
+        match = "Calling-Station-Id = userdata0"
```
And use `match  = "User-Name = login, User-Password = password" here.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -18,7 +18,7 @@ extern "C" STG::Plugin* GetPlugin()
 }
 
-std::vector<std::pair<std::string, AttrValue>> RAD_SETTINGS::ParseAuthAttr(const std::string& value, const std::string& paramName)
+std::vector<std::pair<std::string, AttrValue>> RAD_SETTINGS::ParseAuthAutzAttr(const std::string& value, const std::string& paramName)
```
Let's call it `ParseRules`.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -109,9 +109,9 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
         if (pva != pvi->sections.end() && !pva->value.empty())
         {
-            printfd(__FILE__, "ParseSettings Value of send: '%s'\n", pva->value[0].c_str());
+            printfd(__FILE__, "ParseSettings Value of auth/send: '%s'\n", pva->value[0].c_str());
```
I think, we no longer need this print.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -109,9 +109,9 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
+            m_sendPairs = ParseAuthAutzAttr(pva->value[0], pv.param);
 
             for (const auto& at : m_sendPairs)
                 printfd(__FILE__, "Key: '%s', Value: '%s', Type: %d\n", at.first.c_str(), at.second.value.c_str(), at.second.type);
```
Let's make a function, something like this:
```c++
std::string showRules(const std::vector<std::pair<std::string, AttrValue>>)
{
    // ...
}
```
Which will combine the pairs back to the original form, and use it like this:
```c++
printfd(__FILE__, "auth.match = \"%s\"\n", showRules(m_matchPairs).c_str());
...
```

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -121,9 +121,38 @@ int RAD_SETTINGS::ParseSettings(const ModuleSettings & s)
+        {
+            printfd(__FILE__, "ParseSettings Value of autz/send: '%s'\n", pva->value[0].c_str());
+
+            m_sendPairs = ParseAuthAutzAttr(pva->value[0], pv.param);
```
Here you overwrite the `m_sendPairs`, which contains the value from the 'auth' section. You need additional pair of `m_sendPairs` and `m_matchPairs` for the 'autz' section.

HelenMamontova committed:
Radius. Attributes of auth/send, autz/send, autz/match changed.

HelenMamontova committed:
Radius. ParseSectionsAttr function name changed to ParseRyles.

HelenMamontova committed:
Radius. ParseSectionsAttr function name definition and call changed to ParseRyles.

HelenMamontova committed:
Radius. Extra printfd function call removed in MakeKeyValuePairs
functions.

HelenMamontova committed:
Radius. The ShowRules, GetSendPairsAuth, GetMatchPairsAuth, GetSendPairsAutz, GetMatchPairsAutz
functions declaration added. The m_sendPairs, m_matchPairs class members
removed. The m_sendPairsAuth, m_matchPairsAuth, m_sendPairsAutz, m_matchPairsAutz
class members added.

HelenMamontova committed:
Radius. The ShowRules function definition added and its call in
MakeKeyValuePairs function added.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,127 @@
+        if (at.second.type == AttrValue::Type::PARAM_NAME)
+            strRules.append(at.first + " = " + at.second.value);
+        else
+            strRules.append(at.first + " = " + "'" + at.second.value + "'");
```
You can combine `" = " + "'"` to `" = '"`

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,127 @@
+
+std::string RAD_SETTINGS::ShowRules(const std::vector<std::pair<std::string, AttrValue>> mvector)
+{
+    std::string strRules;
```
Let's just call it `res`, as "result".

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,127 @@
+        else
+            strRules.append(at.first + " = " + "'" + at.second.value + "'");
+
+        if (i != mvector.size() - 1)
```
This `i` counter is really needed once here - at the last element. I propose a slightly modified scheme:
```c++
bool first = true;
for (const auto& at: mvector)
{
    if (first)
        first = false;
    else
        res += ", ";
        
    // The rest of the code...
}
```

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,127 @@
+    return res;
+}
+
+std::string RAD_SETTINGS::ShowRules(const std::vector<std::pair<std::string, AttrValue>> mvector)
```
Missed `&`.

madf replied:
Also let's call it `attributes` or `attrs`. "mvector" says nothing about its nature.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,127 @@
+    return strRules;
+}
+
+void RAD_SETTINGS::MakeKeyValuePairs(const ModuleSettings & s, ParamValue pv, const std::string& paramName)
```
I propose to make it more generic. First, let's make a type:
```c++
struct ASection
{
    using Pairs = std::vector<std::pair<std::string, AttrValue>>;
    Pairs match;
    Pairs send;
};
```

In the `RAD_SETTINGS` use it like this:
```c++
class RAD_SETTINGS
{
    public:
        // ...
        const ASection& getAuth() const { return m_auth; }
        const ASection& getAutz() const { return m_autz; }
        // ...
    private:
        // ...
        ASection m_auth;
        ASection m_autz;
        // ...
};
```

And this function will be:
```c++
ASection parseASection(const std::vector<ParamValue>& conf)
{
    ASection res;
    const auto mit = std::find(conf.begin(), conf.end(), ParamValue("match", {}));
    if (mit != conf.end())
        res.match = ParseRules(mit->value[0], mit->param);
    const auto sit = std::find(conf.begin(), conf.end(), ParamValue("send", {}));
    if (sit != conf.end())
        res.send = ParseRules(sit->value[0], sit->param);
}
```

And then in the `ParseSettings`:
```c++
const auto authIt = std::find(s.moduleParams.begin(), s.moduleParams.end(), ParamValue("auth", {}));
if (authIt != s.moduleParams.end())
    m_auth = parseASection(authIt->sections);
const auto autzIt = std::find(s.moduleParams.begin(), s.moduleParams.end(), ParamValue("autz", {}));
if (autzIt != s.moduleParams.end())
    m_autz = parseASection(autzIt->sections);
printfd(__FILE__, " auth.match = \"%s\"\n", ShowRules(m_auth.match).c_str());
printfd(__FILE__, " auth.send = \"%s\"\n", ShowRules(m_auth.send).c_str());
printfd(__FILE__, " autz.match = \"%s\"\n", ShowRules(m_autz.match).c_str());
printfd(__FILE__, " autz.send = \"%s\"\n", ShowRules(m_autz.send).c_str());
```

HelenMamontova committed:
Radius. The strRules string creating fixed and strRules string renamed to result
in the ShowRules function.

HelenMamontova committed:
Radius. Parameter name changed to attributes and & added to parameter in
the ParseRules function.

HelenMamontova committed:
Radius. The ShowRules function definition changed. Parameter name changed to attributes and & added to parameter in the ShowRules function.

HelenMamontova committed:
Radius. The struct ASection added. The getAuth, getAutz, parseASection
function declaration added. Class member m_auth, m_autz added.
Unnecessary code removed.

HelenMamontova committed:
Radius. The using ASection for alias added. The parseASection function
definition and call added, MakeKeyValuePairs function removed,
ParseSettings function changed. Extra printfd function added.

HelenMamontova committed:
Radius. The return added to parseASection function. Extra printfd
functions removed.

HelenMamontova committed:
Radius. GIT_TAG changed to 1.1.3 in the ExternalProject_Add for
async-radius.

HelenMamontova committed:
Radius. The m_ioService class member replaced by m_ioContext in class
Radius.

HelenMamontova committed:
Radius. The m_ioService class member replaced by m_ioContext in the
function Run.

HelenMamontova committed:
Radius. Parameter io_service replaced by io_context in the constructor Server.

HelenMamontova committed:
Radius. Parameter io_service replaced by io_context in the constructor Server definition.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -47,27 +77,27 @@ namespace STG
             uint16_t GetStopPosition() const override { return 0; }
 
-            int SendMessage(const Message & msg, uint32_t ip) const override { return 0; }
+            int SendMessage(const Message& msg, uint32_t ip) const override { return 0; }
```
```
/tmp/stg/projects/stargazer/plugins/other/radius/radius.h: In member function ‘virtual int STG::RADIUS::SendMessage(const STG::Message&, uint32_t) const’:
/tmp/stg/projects/stargazer/plugins/other/radius/radius.h:95:44: warning: unused parameter ‘msg’ [-Wunused-parameter]
   95 |             int SendMessage(const Message& msg, uint32_t ip) const override { return 0; }
      |                             ~~~~~~~~~~~~~~~^~~
/tmp/stg/projects/stargazer/plugins/other/radius/radius.h:95:58: warning: unused parameter ‘ip’ [-Wunused-parameter]
   95 |             int SendMessage(const Message& msg, uint32_t ip) const override { return 0; }

```

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,104 @@
+std::vector<std::pair<std::string, AttrValue>> RAD_SETTINGS::ParseRules(const std::string& value, const std::string& paramName)
+{
+    using tokenizer =  boost::tokenizer<boost::char_separator<char>>;
+    boost::char_separator<char> sep(",");
```
```
/tmp/stg/projects/stargazer/plugins/other/radius/radius.cpp:25:5: warning: variable 'sep' of type 'boost::char_separator<char>' can be declared 'const' [misc-const-correctness]
   25 |     boost::char_separator<char> sep(",");
      |     ^
      |                                 const 
```

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,104 @@
+    using tokenizer =  boost::tokenizer<boost::char_separator<char>>;
+    boost::char_separator<char> sep(",");
+
+    tokenizer tokens(value, sep);
```
```
/tmp/stg/projects/stargazer/plugins/other/radius/radius.cpp:27:5: warning: variable 'tokens' of type 'tokenizer' (aka 'tokenizer<boost::char_separator<char>>') can be declared 'const' [misc-const-correctness]
   27 |     tokenizer tokens(value, sep);
      |     ^
      |               const
```

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,104 @@
+    std::vector<std::pair<std::string, AttrValue>> res;
+    for (const auto& token : tokens)
+    {
+        boost::char_separator<char> sp(" =");
```
```
/tmp/stg/projects/stargazer/plugins/other/radius/radius.cpp:32:9: warning: variable 'sp' of type 'boost::char_separator<char>' can be declared 'const' [misc-const-correctness]
   32 |         boost::char_separator<char> sp(" =");
      |         ^
      |                                     const
```

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,104 @@
+    for (const auto& token : tokens)
+    {
+        boost::char_separator<char> sp(" =");
+        tokenizer tok(token, sp);
```
```
/tmp/stg/projects/stargazer/plugins/other/radius/radius.cpp:33:9: warning: variable 'tok' of type 'tokenizer' (aka 'tokenizer<boost::char_separator<char>>') can be declared 'const' [misc-const-correctness]
   33 |         tokenizer tok(token, sp);
      |         ^
      |                   const
```

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,104 @@
+    return res;
+}
+
+std::string RAD_SETTINGS::ShowRules(const std::vector<std::pair<std::string, AttrValue>>& attributes)
```
This function does not use anything from `RAD_SETTINGS`. Consider making it a free function.

HelenMamontova committed:
Radius. Parameters msg, ip commented out in the SendMessage function.

HelenMamontova committed:
Radius. Variable 'sep' declared 'const' in ParseRules function.

HelenMamontova committed:
Radius. Variables 'tokens', 'sp', 'tok' declared 'const' in ParseRules function.

HelenMamontova committed:
Radius. ShowRules function declaration removed from class RAD_SETTINGS.

HelenMamontova committed:
Radius. ShowRules function definition is made as a free function.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -19,18 +19,37 @@ extern "C" STG::Plugin* GetPlugin()
     return &plugin;
 }
 
+std::string ShowRules(const std::vector<std::pair<std::string, AttrValue>>& attributes)
```
Please put it into the anonymous namespace.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -19,18 +19,37 @@ extern "C" STG::Plugin* GetPlugin()
+    bool first = true;
+    for (const auto& at : attributes)
+    {
+        if (first)
```
I think we can re-work this coma placement code. What do you think about this:
```c++
if (!result.empty())
    result += ", ";
```

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -47,27 +76,27 @@ namespace STG
             RADIUS& operator=(const RADIUS&) = delete;
 
             void SetUsers(Users* u) { m_users = u; }
```
This function must be marked `override`.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -1,22 +1,104 @@
 
+#include <vector>
+#include <utility>
 #include <iterator>
```
Add `#include <algorithm>` - you use `std::find` from there.

HelenMamontova committed:
Radius. ShowRules function definition is placed in the anonymous namespace.

HelenMamontova committed:
Radius. The code for determining the first element of vector changed in the function ShowRules.

HelenMamontova committed:
Radius. Function SetUsers declaration is made as override in the class
Radius.

HelenMamontova committed:
Radius. Header file <algorithm> added.
